### PR TITLE
Handling sending message error

### PR DESCRIPTION
**Pull request by apereverzin**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-1206

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java [hunk 1]
```diff
@@ -323,22 +324,38 @@ private void handleInvalidFileError(
...
+            ErrorMapping.getFor(cause.getClass()),
+            (errorCode) -> trySendErrorMessageToQueueAndMoveFileToRejectedContainer(
+                zipFilename,
+                containerName,
+                cause, eventId,
+                leaseId,
+                errorCode
+            ),
+            () -> blobManager.tryMoveFileToRejectedContainer(zipFilename, containerName, leaseId)
+        );
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:337]: It should be more sth like this:
```
Optionals.ifPresentOrElse(
...
{
   send zip file
   move to rejected container when no error so far
},
() -> throw exception because we don't have necessary error code mapping (this is an existing bug)
```

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java [hunk 2]
```diff
@@ -323,22 +324,40 @@ private void handleInvalidFileError(
...
+        try {
+            sendErrorMessageToQueue(zipFilename, containerName, eventId, errorCode, cause);
+            blobManager.tryMoveFileToRejectedContainer(zipFilename, containerName, leaseId);
+        } catch (Exception exc) {
+            log.error(
+                "Error sending notification to the queue."
+                    + "File name: " + zipFilename + " "
+                    + "Container: " + containerName,
+                exc
+            );
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:359]: All we need to do here is not catch the exception and let the process fail.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java [hunk 3]
```diff
@@ -323,22 +324,41 @@ private void handleInvalidFileError(
...
+            (errorCode) -> trySendErrorMessageAndMoveFileToRejectedContainer(
+                zipFilename,
+                containerName,
+                cause, eventId,
+                leaseId,
+                errorCode
+            ),
+            () -> {
+                throw new ConfigurationException("Error code mapping not found for " + cause.getClass().getName());
+            }
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:338]: unrelated

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:338]: This is what Leszek asked to change

Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:323]: I asked for this after some changes were made in this PR, resulting in the lack of failure in case of misconfiguration.

All we need for this story, is not catching the exception.

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:323]: ok, I will revert this change

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java [hunk 4]
```diff
@@ -308,22 +309,39 @@ private void handleInvalidFileError(
...
+        Optionals.ifPresentOrElse(
+            ErrorMapping.getFor(cause.getClass()),
+            (errorCode) -> trySendErrorMessageAndMoveFileToRejectedContainer(
+                zipFilename,
+                containerName,
+                cause, eventId,
+                leaseId,
+                errorCode
+            ),
+            () -> blobManager.tryMoveFileToRejectedContainer(zipFilename, containerName, leaseId)
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:321]: > I asked for this after some changes were made in this PR, resulting in the lack of failure in case of misconfiguration.
Sorry, I was wrong - the problem was there even before your changes. We should throw an exception here. The most important change in this PR is to not catch exceptions thrown when we fail to send message.

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java]: Oh! I have thought you meant to raise another task for this

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java [hunk 5]
```diff
@@ -308,22 +309,41 @@ private void handleInvalidFileError(
...
+            sendErrorMessageToQueue(zipFilename, containerName, eventId, errorCode, cause);
+            blobManager.tryMoveFileToRejectedContainer(zipFilename, containerName, leaseId);
+        } catch (Exception exc) {
+            log.error(
+                "Error sending notification to the queue."
+                    + "File name: " + zipFilename + " "
+                    + "Container: " + containerName,
+                exc
+            );
+            throw exc;
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:345]: I missed this change.

We should either not catch this exception at all or throw a new exception with our custom message. We can't log an error and then throw an exception, because that will result in two error log entries in App Insights, representing the same thing.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java [hunk 6]
```diff
@@ -308,22 +310,38 @@ private void handleInvalidFileError(
...
+        String leaseId,
+        ErrorCode errorCode
+    ) {
+        try {
+            sendErrorMessageToQueue(zipFilename, containerName, eventId, errorCode, cause);
+            blobManager.tryMoveFileToRejectedContainer(zipFilename, containerName, leaseId);
+        } catch (Exception exc) {
+            final String msg = "Error sending notification to the queue."
+                + "File name: " + zipFilename + " "
+                + "Container: " + containerName;
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:342]: The message is about sending notification to the queue, but the try part is about more than that. Perhaps we should remove the call to blobManager from this method.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java [hunk 7]
```diff
@@ -308,22 +310,38 @@ private void handleInvalidFileError(
...
+                errorCode
+            ),
+            () -> {
+                throw new ConfigurationException("Error code mapping not found for " + cause.getClass().getName());
+            }
+        );
+    }
 
-        blobManager.tryMoveFileToRejectedContainer(zipFilename, containerName, leaseId);
+    private void trySendErrorMessageAndMoveFileToRejectedContainer(
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:328]: Now that this method throws exception on failure, best to remove `try` prefix. And if we remove the call to blob manager from here, ale the name part related with that bit.

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForDisabledService.java [hunk 8]
```diff
@@ -75,4 +81,36 @@ public void should_reject_file_when_service_is_disabled() throws Exception {
         errorWasSent(SAMPLE_ZIP_FILE_NAME, ErrorCode.ERR_SERVICE_DISABLED);
     }
 
+    @Test
+    public void should_not_move_file_if_send_message_throws_exception_when_service_is_disabled() throws Exception {
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForDisabledService.java:85]: How about something more general, e.g. `should_abandon_rejecting_file_when_sending_error_notification_fails`? The reasons of why we're trying to reject the file is irrelevant here.

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForDisabledService.java [hunk 9]
```diff
@@ -75,4 +81,36 @@ public void should_reject_file_when_service_is_disabled() throws Exception {
...
+        // given
+        byte[] zipBytes = zipDir("zipcontents/ok");
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForDisabledService.java:87]: Could we maybe use an invalid file instead of using the assumption that the service is disabled? The test code doesn't make the service disabled - it relies on some external condition. It would be best if the test itself could say "look, this is the bit that will cause the rejection".

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForDisabledService.java (hunk 8 above)
Comment (apereverzin) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForDisabledService.java]: Done. And new test class created

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForDisabledService.java (hunk 9 above)
Comment (apereverzin) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForDisabledService.java]: Created new test class BlobProcessorTaskTestForFailingNotification

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java (hunk 7 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java]: Done

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java (hunk 6 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:338]: Done

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java [hunk 10]
```diff
@@ -304,24 +306,40 @@ private void handleInvalidFileError(
...
-
+        Optionals.ifPresentOrElse(
+            ErrorMapping.getFor(cause.getClass()),
+            (errorCode) -> sendErrorMessage(
+                zipFilename,
+                containerName,
+                cause, eventId,
+                leaseId,
+                errorCode
+            ),
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:317]: Suggestion: it would be best to put `blobManager.tryMoveFileToRejectedContainer(...` here. The code will still work as it is, but it would be clearer that we only proceed with both actions if the mapping is there.

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailingNotification.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+import static uk.gov.hmcts.reform.bulkscanprocessor.model.common.Event.ZIPFILE_PROCESSING_STARTED;
+
+@IntegrationTest
+@RunWith(SpringRunner.class)
+@TestPropertySource(properties = {
+    "containers.mappings[0].container=bulkscan",
+    "containers.mappings[0].jurisdiction=BULKSCAN",
+    "containers.mappings[0].poBox=BULKSCANPO",
+    "containers.mappings[0].paymentsEnabled=false",
+    "containers.mappings[0].enabled=false"
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailingNotification.java:28]: These properties seem irrelevant to what we're testing here. Can they be removed?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailingNotification.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+            zipFileProcessor,
+            containerMappings,
+            ocrValidator,
+            serviceBusHelper,
+            paymentsEnabled
+        );
+    }
+
+    @Test
+    public void should_abandon_rejected_file_when_sending_error_notification_fails() throws Exception {
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailingNotification.java:48]: The file is not rejected yet. How about `should_abandon_processing_invalid_file...`?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailingNotification.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+@IntegrationTest
+@RunWith(SpringRunner.class)
+@TestPropertySource(properties = {
+    "containers.mappings[0].container=bulkscan",
+    "containers.mappings[0].jurisdiction=BULKSCAN",
+    "containers.mappings[0].poBox=BULKSCANPO",
+    "containers.mappings[0].paymentsEnabled=false",
+    "containers.mappings[0].enabled=false"
+})
+public class BlobProcessorTaskTestForFailingNotification extends ProcessorTestSuite<BlobProcessorTask> {
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailingNotification.java:30]: By convention, `Test` should go to the end of the test class name.

Comment (apereverzin) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailingNotification.java:30]: There are quite a lot of test classes in this package named BlobProcessorTaskTestFor..., and this approach makes sense in my opinion

Comment (doncem) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailingNotification.java:30]: we have few of those. i guess shouldn't be harmful here either

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailingNotification.java (hunk 12 above)
Comment (apereverzin) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailingNotification.java:48]: ok

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailingNotification.java (hunk 13 above)
Comment (lgonczar) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTaskTestForFailingNotification.java:22]: Aaa, I see, and all of them are for BlobProcessorTask. Fair enough. Might need to rename them all on one go later.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java [hunk 14]
```diff
@@ -304,22 +306,40 @@ private void handleInvalidFileError(
...
-                    );
-                }
-            });
+        Optionals.ifPresentOrElse(
+            ErrorMapping.getFor(cause.getClass()),
+            (errorCode) -> {
+                sendErrorMessage(
+                    zipFilename,
+                    containerName,
+                    cause, eventId,
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:315]: Rafał won't like this line

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java [hunk 15]
```diff
@@ -304,22 +306,40 @@ private void handleInvalidFileError(
...
+        String containerName,
+        Exception cause,
+        Long eventId,
+        String leaseId,
+        ErrorCode errorCode
+    ) {
+        try {
+            sendErrorMessageToQueue(zipFilename, containerName, eventId, errorCode, cause);
+        } catch (Exception exc) {
+            final String msg = "Error sending notification to the queue."
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:338]: Sorry to be requesting so many changes for this PR. It would be best for this message to be sth like `Error sending error notification...`, because we have notification by which we understand the ones we send to the orchestrator.

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java:338]: No worries at all.
